components/brave_rewards/core/engine/hash_prefix_store.cc

```diff
@@ -0,0 +1,169 @@
+/* Copyright (c) 2025 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_rewards/core/engine/hash_prefix_store.h"
+
+#include <algorithm>
+
+#include "base/containers/span.h"
+#include "base/containers/span_reader.h"
+#include "base/files/file.h"
+#include "base/memory/raw_span.h"
+#include "base/numerics/byte_conversions.h"
+#include "brave/components/brave_rewards/core/engine/hash_prefix_iterator.h"
+#include "crypto/sha2.h"
+
+namespace brave_rewards {
+
+namespace {
+
+constexpr uint32_t kFileVersion = 1;
+constexpr uint32_t kMinPrefixSize = 4;
+constexpr uint32_t kMaxPrefixSize = 32;
+
+struct FileParts {
+  uint32_t version = 0;
+  uint32_t prefix_size = 0;
+  base::raw_span<const uint8_t> prefixes;
+};
+
+bool IsValidFile(const FileParts& parts) {
+  if (parts.version != kFileVersion) {
+    return false;
+  }
+  if (parts.prefix_size < kMinPrefixSize ||
+      parts.prefix_size > kMaxPrefixSize) {
+    return false;
+  }
+  if (parts.prefixes.size() % parts.prefix_size != 0) {
+    return false;
+  }
+  return true;
+}
+
+std::optional<FileParts> ParseFile(base::span<const uint8_t> bytes) {
+  FileParts parts;
+  base::SpanReader reader(bytes);
+  if (!reader.ReadU32LittleEndian(parts.version)) {
+    return std::nullopt;
+  }
+  if (!reader.ReadU32LittleEndian(parts.prefix_size)) {
+    return std::nullopt;
+  }
+  parts.prefixes = reader.remaining_span();
```

Comment:
I believe we need to verify size here. Otherwise we can get an overflow in case the file is broken (i.e. with a sudden shutdown)

Reply:
See `IsValidFile` for span-length validation.

Reply:
Good catch! Added a prefix count to the header. The file now looks like:

```
[version:u32] [prefix_size:u32] [prefix_count:u64] [data...]
```